File: taskm-sdk-java/src/main/java/com/taskm/sdk/Environment.java
package com.taskm.sdk;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Utility class for accessing environment variables with type conversion.
 *
 * <p>This class provides convenient methods to read environment variables
 * with automatic type conversion and default values.
 *
 * <h3>Common environment variables in HS-TASKM:</h3>
 * <ul>
 *   <li>TASK_ID: Task execution ID</li>
 *   <li>LOG_TYPE: Log type (strategy, plugin, listener)</li>
 *   <li>PLUGIN_ENDPOINT: Plugin instance HTTP endpoint</li>
 *   <li>LISTENER_ENDPOINT: Listener instance HTTP endpoint</li>
 * </ul>
 *
 * <h3>Example usage:</h3>
 * <pre>{@code
 * import com.taskm.sdk.Environment;
 *
 * // Get with type conversion and default value
 * int taskId = Environment.getInt("TASK_ID", 0);
 * String logType = Environment.get("LOG_TYPE", "strategy");
 * boolean enabled = Environment.getBoolean("ENABLED", false);
 *
 * // Get required variable (throws exception if not set)
 * String apiKey = Environment.getRequired("API_KEY");
 *
 * // Get all variables with prefix
 * Map<String, String> taskmVars = Environment.getAll("TASKM_");
 *
 * // Check if variable exists
 * if (Environment.exists("DEBUG")) {
 *     // ...
 * }
 * }</pre>
 *
 * @since 1.0.0
 */
public final class Environment {

    // Private constructor to prevent instantiation
    private Environment() {
        throw new UnsupportedOperationException("Utility class cannot be instantiated");
    }

    /**
     * Get an environment variable as a string.
     *
     * @param name the environment variable name
     * @param defaultValue the default value if variable is not set
     * @return the environment variable value, or the default value if not set
     */
    public static String get(String name, String defaultValue) {
        String value = System.getenv(name);
        return value != null ? value : defaultValue;
    }

    /**
     * Get an environment variable as a string.
     *
     * @param name the environment variable name
     * @return the environment variable value, or null if not set
     */
    public static String get(String name) {
        return System.getenv(name);
    }

    /**
     * Get an environment variable as an integer.
     *
     * @param name the environment variable name
     * @param defaultValue the default value if variable is not set
     * @return the environment variable value as an integer, or the default value
     * @throws IllegalArgumentException if the value cannot be converted to an integer
     */
    public static int getInt(String name, int defaultValue) {
        String value = System.getenv(name);
        if (value == null) {
            return defaultValue;
        }
        try {
            return Integer.parseInt(value.trim());
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                String.format("Environment variable '%s' value '%s' cannot be converted to int",
                    name, value), e);
        }
    }

    /**
     * Get an environment variable as a long.
     *
     * @param name the environment variable name
     * @param defaultValue the default value if variable is not set
     * @return the environment variable value as a long, or the default value
     * @throws IllegalArgumentException if the value cannot be converted to a long
     */
    public static long getLong(String name, long defaultValue) {
        String value = System.getenv(name);
        if (value == null) {
            return defaultValue;
        }
        try {
            return Long.parseLong(value.trim());
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                String.format("Environment variable '%s' value '%s' cannot be converted to long",
                    name, value), e);
        }
    }

    /**
     * Get an environment variable as a float.
     *
     * @param name the environment variable name
     * @param defaultValue the default value if variable is not set
     * @return the environment variable value as a float, or the default value
     * @throws IllegalArgumentException if the value cannot be converted to a float
     */
    public static float getFloat(String name, float defaultValue) {
        String value = System.getenv(name);
        if (value == null) {
            return defaultValue;
        }
        try {
            return Float.parseFloat(value.trim());
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                String.format("Environment variable '%s' value '%s' cannot be converted to float",
                    name, value), e);
        }
    }

    /**
     * Get an environment variable as a double.
     *
     * @param name the environment variable name
     * @param defaultValue the default value if variable is not set
     * @return the environment variable value as a double, or the default value
     * @throws IllegalArgumentException if the value cannot be converted to a double
     */
    public static double getDouble(String name, double defaultValue) {
        String value = System.getenv(name);
        if (value == null) {
            return defaultValue;
        }
        try {
            return Double.parseDouble(value.trim());
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                String.format("Environment variable '%s' value '%s' cannot be converted to double",
                    name, value), e);
        }
    }

    /**
     * Get an environment variable as a boolean.
     *
     * <p>Accepts: true/false, 1/0, yes/no, on/off (case-insensitive)
     *
     * @param name the environment variable name
     * @param defaultValue the default value if variable is not set
     * @return the environment variable value as a boolean, or the default value
     * @throws IllegalArgumentException if the value cannot be converted to a boolean
     */
    public static boolean getBoolean(String name, boolean defaultValue) {
        String value = System.getenv(name);
        if (value == null) {
            return defaultValue;
        }

        String normalized = value.trim().toLowerCase();
        switch (normalized) {
            case "true":
            case "1":
            case "yes":
            case "on":
            case "t":
            case "y":
                return true;
            case "false":
            case "0":
            case "no":
            case "off":
            case "f":
            case "n":
            case "":
                return false;
            default:
                throw new IllegalArgumentException(
                    String.format("Environment variable '%s' value '%s' cannot be converted to boolean",
                        name, value));
        }
    }

    /**
     * Get a required environment variable.
     *
     * @param name the environment variable name
     * @return the environment variable value
     * @throws IllegalArgumentException if the environment variable is not set
     */
    public static String getRequired(String name) {
        String value = System.getenv(name);
        if (value == null) {
            throw new IllegalArgumentException(
                String.format("Required environment variable '%s' is not set", name));
        }
        return value;
    }

    /**
     * Get a required environment variable as an integer.
     *
     * @param name the environment variable name
     * @return the environment variable value as an integer
     * @throws IllegalArgumentException if the environment variable is not set
     *          or cannot be converted to an integer
     */
    public static int getRequiredInt(String name) {
        String value = getRequired(name);
        try {
            return Integer.parseInt(value.trim());
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                String.format("Environment variable '%s' value '%s' cannot be converted to int",
                    name, value), e);
        }
    }

    /**
     * Get a required environment variable as a long.
     *
     * @param name the environment variable name
     * @return the environment variable value as a long
     * @throws IllegalArgumentException if the environment variable is not set
     *          or cannot be converted to a long
     */
    public static long getRequiredLong(String name) {
        String value = getRequired(name);
        try {
            return Long.parseLong(value.trim());
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                String.format("Environment variable '%s' value '%s' cannot be converted to long",
                    name, value), e);
        }
    }

    /**
     * Get a required environment variable as a double.
     *
     * @param name the environment variable name
     * @return the environment variable value as a double
     * @throws IllegalArgumentException if the environment variable is not set
     *          or cannot be converted to a double
     */
    public static double getRequiredDouble(String name) {
        String value = getRequired(name);
        try {
            return Double.parseDouble(value.trim());
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException(
                String.format("Environment variable '%s' value '%s' cannot be converted to double",
                    name, value), e);
        }
    }

    /**
     * Get a required environment variable as a boolean.
     *
     * @param name the environment variable name
     * @return the environment variable value as a boolean
     * @throws IllegalArgumentException if the environment variable is not set
     *          or cannot be converted to a boolean
     */
    public static boolean getRequiredBoolean(String name) {
        String value = getRequired(name);
        String normalized = value.trim().toLowerCase();
        switch (normalized) {
            case "true":
            case "1":
            case "yes":
            case "on":
            case "t":
            case "y":
                return true;
            case "false":
            case "0":
            case "no":
            case "off":
            case "f":
            case "n":
            case "":
                return false;
            default:
                throw new IllegalArgumentException(
                    String.format("Environment variable '%s' value '%s' cannot be converted to boolean",
                        name, value));
        }
    }

    /**
     * Get all environment variables.
     *
     * @return a map of all environment variable names and values
     */
    public static Map<String, String> getAll() {
        return new HashMap<>(System.getenv());
    }

    /**
     * Get all environment variables with a specific prefix.
     *
     * @param prefix the prefix to filter variables (e.g., "TASKM_")
     * @return a map of environment variable names and values with the specified prefix
     */
    public static Map<String, String> getAll(String prefix) {
        if (prefix == null || prefix.isEmpty()) {
            return getAll();
        }

        return System.getenv().entrySet().stream()
            .filter(entry -> entry.getKey().startsWith(prefix))
            .collect(Collectors.toMap(
                Map.Entry::getKey,
                Map.Entry::getValue,
                (existing, replacement) -> existing,
                HashMap::new
            ));
    }

    /**
     * Check if an environment variable is set.
     *
     * @param name the environment variable name
     * @return true if the variable is set, false otherwise
     */
    public static boolean exists(String name) {
        return System.getenv(name) != null;
    }

    /**
     * Check if any of the specified environment variables are set.
     *
     * @param names the environment variable names to check
     * @return true if any of the variables are set, false otherwise
     */
    public static boolean existsAny(String... names) {
        if (names == null || names.length == 0) {
            return false;
        }
        for (String name : names) {
            if (exists(name)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Check if all of the specified environment variables are set.
     *
     * @param names the environment variable names to check
     * @return true if all of the variables are set, false otherwise
     */
    public static boolean existsAll(String... names) {
        if (names == null || names.length == 0) {
            return true;
        }
        for (String name : names) {
            if (!exists(name)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Get environment variable value split by delimiter.
     *
     * <p>Useful for comma-separated or colon-separated values.
     *
     * @param name the environment variable name
     * @param delimiter the delimiter to split on (e.g., ",", ":")
     * @return a list of split values, or an empty list if the variable is not set
     */
    public static List<String> getList(String name, String delimiter) {
        String value = System.getenv(name);
        if (value == null || value.isEmpty()) {
            return new ArrayList<>();
        }

        return Arrays.stream(value.split(delimiter))
            .map(String::trim)
            .filter(s -> !s.isEmpty())
            .collect(Collectors.toList());
    }

    /**
     * Get environment variable value split by comma.
     *
     * @param name the environment variable name
     * @return a list of comma-separated values, or an empty list if the variable is not set
     */
    public static List<String> getCommaSeparatedList(String name) {
        return getList(name, ",");
    }
}

File: taskm-sdk-py/taskm_sdk/environment.py
"""
Environment variable utilities for HS-TASKM SDK.

This module provides convenient methods for accessing environment variables
with type conversion and default value support.
"""

import os
from typing import Any, Dict, Optional, TypeVar, Type, Union

T = TypeVar('T', str, int, float, bool)


class Environment:
    """
    Utility class for accessing environment variables with type conversion.

    This class provides convenient methods to read environment variables
    with automatic type conversion and default values.

    Common environment variables in HS-TASKM:
    - TASK_ID: Task execution ID
    - LOG_TYPE: Log type (strategy, plugin, listener)
    - PLUGIN_ENDPOINT: Plugin instance HTTP endpoint
    - LISTENER_ENDPOINT: Listener instance HTTP endpoint

    Example:
        >>> from taskm_sdk import Environment
        >>> task_id = Environment.get("TASK_ID", default=0, type=int)
        >>> log_type = Environment.get("LOG_TYPE", default="strategy")
        >>> enabled = Environment.get("ENABLED", type=bool)
        >>> all_vars = Environment.get_all()
    """

    @staticmethod
    def get(
        name: str,
        default: Optional[T] = None,
        type: Optional[Type[T]] = None
    ) -> Optional[T]:
        """
        Get an environment variable with optional type conversion and default value.

        Args:
            name: Environment variable name
            default: Default value if variable is not set
            type: Type to convert the value to (str, int, float, bool)

        Returns:
            The environment variable value, converted to the specified type,
            or the default value if not set.

        Raises:
            ValueError: If type conversion fails
            TypeError: If an unsupported type is specified

        Example:
            >>> Environment.get("TASK_ID", default=0, type=int)
            123
            >>> Environment.get("LOG_TYPE", default="strategy")
            'strategy'
            >>> Environment.get("ENABLED", type=bool)
            True
            >>> Environment.get("PORT", default=8080, type=int)
            8080
        """
        value = os.getenv(name)

        if value is None:
            return default

        if type is None:
            return value

        # Type conversion
        if type == str:
            return value

        elif type == int:
            try:
                return int(value)
            except ValueError:
                raise ValueError(
                    f"Environment variable '{name}' value '{value}' "
                    f"cannot be converted to int"
                )

        elif type == float:
            try:
                return float(value)
            except ValueError:
                raise ValueError(
                    f"Environment variable '{name}' value '{value}' "
                    f"cannot be converted to float"
                )

        elif type == bool:
            # Handle various boolean representations
            if isinstance(value, bool):
                return value

            normalized = value.strip().lower()
            if normalized in ('true', '1', 'yes', 'on', 't', 'y'):
                return True
            elif normalized in ('false', '0', 'no', 'off', 'f', 'n', ''):
                return False
            else:
                raise ValueError(
                    f"Environment variable '{name}' value '{value}' "
                    f"cannot be converted to bool"
                )

        else:
            raise TypeError(
                f"Unsupported type '{type}'. "
                f"Supported types: str, int, float, bool"
            )

    @staticmethod
    def get_int(name: str, default: Optional[int] = None) -> Optional[int]:
        """
        Get an environment variable as an integer.

        Args:
            name: Environment variable name
            default: Default value if variable is not set

        Returns:
            Integer value or default

        Example:
            >>> Environment.get_int("PORT", 8080)
            8080
            >>> Environment.get_int("TASK_ID")
            123
        """
        return Environment.get(name, default=default, type=int)

    @staticmethod
    def get_float(name: str, default: Optional[float] = None) -> Optional[float]:
        """
        Get an environment variable as a float.

        Args:
            name: Environment variable name
            default: Default value if variable is not set

        Returns:
            Float value or default

        Example:
            >>> Environment.get_float("PRICE", 0.0)
            100.5
        """
        return Environment.get(name, default=default, type=float)

    @staticmethod
    def get_bool(name: str, default: Optional[bool] = None) -> Optional[bool]:
        """
        Get an environment variable as a boolean.

        Accepts: true/false, 1/0, yes/no, on/off, t/f, y/n (case-insensitive)

        Args:
            name: Environment variable name
            default: Default value if variable is not set

        Returns:
            Boolean value or default

        Example:
            >>> Environment.get_bool("ENABLED", False)
            True
            >>> Environment.get_bool("DEBUG")
            False
        """
        return Environment.get(name, default=default, type=bool)

    @staticmethod
    def get_string(name: str, default: Optional[str] = None) -> Optional[str]:
        """
        Get an environment variable as a string.

        Args:
            name: Environment variable name
            default: Default value if variable is not set

        Returns:
            String value or default

        Example:
            >>> Environment.get_string("LOG_TYPE", "strategy")
            'strategy'
        """
        return Environment.get(name, default=default, type=str)

    @staticmethod
    def get_required(name: str, type: Optional[Type[T]] = None) -> T:
        """
        Get a required environment variable.

        Raises an exception if the variable is not set.

        Args:
            name: Environment variable name
            type: Type to convert the value to (str, int, float, bool)

        Returns:
            The environment variable value, converted to the specified type

        Raises:
            ValueError: If the environment variable is not set or type conversion fails

        Example:
            >>> Environment.get_required("TASK_ID", type=int)
            123
            >>> Environment.get_required("API_KEY")
            'secret-key'
        """
        value = os.getenv(name)
        if value is None:
            raise ValueError(f"Required environment variable '{name}' is not set")

        return Environment.get(name, type=type)

    @staticmethod
    def get_all(prefix: Optional[str] = None) -> Dict[str, str]:
        """
        Get all environment variables, optionally filtered by prefix.

        Args:
            prefix: Optional prefix to filter variables (e.g., "TASKM_")

        Returns:
            Dictionary of environment variable names and values

        Example:
            >>> Environment.get_all()
            {'TASK_ID': '123', 'LOG_TYPE': 'strategy', ...}
            >>> Environment.get_all("TASKM_")
            {'TASKM_API_KEY': 'xxx', 'TASKM_TIMEOUT': '30'}
        """
        if prefix:
            return {
                key: value
                for key, value in os.environ.items()
                if key.startswith(prefix)
            }
        return dict(os.environ)

    @staticmethod
    def exists(name: str) -> bool:
        """
        Check if an environment variable is set.

        Args:
            name: Environment variable name

        Returns:
            True if the variable is set, False otherwise

        Example:
            >>> Environment.exists("TASK_ID")
            True
            >>> Environment.exists("NON_EXISTENT")
            False
        """
        return name in os.environ
